### PR TITLE
Visualization for Deconvolutional Auto-encoder & test code for deconvolution

### PR DESCRIPTION
**Pull request by Wangyida**
### 1. Visualization for Deconvolutional Auto-encoder.
Example for MNIST dataset, A deconvolutional auto-encoder which is different from mlp. It is related to the deconvolutional semantic segmentation codes with parameters from Caffe.

### 2. Test code for deconvolution algorithm.
It includes forward and backward checking for deconvolutional layer in tiny_cnn. I use the 3*3 weight and 2*2 patch for testing

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -178,7 +178,7 @@ ENDIF()
 
 
 IF(BUILD_TESTS)
-    ADD_EXECUTABLE(tiny_cnn_test test/test.cpp test/test_no_duplicate_symbols.cpp  ${tiny_cnn_hdrs} ${tiny_cnn_test_headers})
+    ADD_EXECUTABLE(tiny_cnn_test test/test.cpp ${tiny_cnn_hdrs} ${tiny_cnn_test_headers})
 ENDIF()
```
Comment (edgarriba) [inline CMakeLists.txt]: why do you remove it? It was added in this PR https://github.com/nyanp/tiny-cnn/pull/171

**Comment by edgarriba** [comment]
@Wangyida You cannot add picotest.h, it must remain as a git submodule.Besides, as I see it is an empty file

File: CMakeLists.txt (hunk 1 above)
Comment (Wangyida) [inline CMakeLists.txt]: @edgarriba Thank you, I will correct these areas, test/test_no_duplicate_symbols.cpp will be restored.

File: examples/deconv/visual.cpp [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT 
+    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS 
+    SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+#include <iostream>
+#include <opencv2/opencv.hpp>
+/*#include <opencv2/imgproc.hpp>
+#include <opencv2/imgcodecs.hpp>
+#include <opencv2/highgui.hpp>*/
+#include "tiny_cnn/tiny_cnn.h"
```
Comment (edgarriba) [inline examples/deconv/visual.cpp]: Please remove commented code

File: test/test_deconvolutional_layer.h [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+*/
+#pragma once
+#include "picotest/picotest.h"
+#include "testhelper.h"
+#include "tiny_cnn/tiny_cnn.h"
+
+namespace tiny_cnn {
+
+TEST(convolutional, fprop) {
+    typedef network<sequential> CNN;
```
Comment (edgarriba) [inline test/test_deconvolutional_layer.h]: name it as `TEST(deconvolutional, fprop)`

File: test/test_deconvolutional_layer.h [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        EXPECT_NEAR(0.5249791, out[11], 1E-5);
+        EXPECT_NEAR(0.5374298, out[12], 1E-5);
+        EXPECT_NEAR(0.3543437, out[13], 1E-5);
+        EXPECT_NEAR(0.5374298, out[14], 1E-5);
+        EXPECT_NEAR(0.5000000, out[15], 1E-5);
+    }
+}
+
+TEST(convolutional, gradient_check) { // tanh - mse
+    network<sequential> nn;
```
Comment (edgarriba) [inline test/test_deconvolutional_layer.h]: Same here, and so on: `TEST(deconvolutional, gradient_check)`

**Comment by Wangyida** [comment]
@edgarriba OK, I will correct these details. You haven't check it right now as I find some macro definition and backward pass problems.

File: CMakeLists.txt [hunk 5]
```diff
@@ -11,7 +11,7 @@
 CMAKE_MINIMUM_REQUIRED(VERSION 2.6)
 
 PROJECT(tiny_cnn)
-set(PROJECT_VERSION "0.0.1")
+set(PROJECT_VERSION "0.1.0")
 string(REGEX MATCHALL "[0-9]" PROJECT_VERSION_PARTS "${PROJECT_VERSION}")
```
Comment (edgarriba) [inline CMakeLists.txt]: Did you rebase master branch?

**Comment by Wangyida** [comment]
No, I rebase this codes still on feat/generic-computational-graph, is the version should be remained as 0.0.1 except for master branch?

File: CMakeLists.txt (hunk 5 above)
Comment (Wangyida) [inline CMakeLists.txt]: No, I rebase this codes still on feat/generic-computational-graph, is the version should be remained as 0.0.1 except for master branch?

Comment (edgarriba) [inline CMakeLists.txt]: @nyanp Well yes, you are right, there's no coordination between branches. We discussed it in other thread. https://github.com/nyanp/tiny-cnn/pull/181

Comment (Wangyida) [inline CMakeLists.txt]: OK, Thanks, I will push again after I fix the back forward pass tests.

**Comment by edgarriba** [comment]
@Wangyida you need to computational graph branch

**Comment by Wangyida** [comment]
@edgarriba Thanks, I noticed that you PR is merged now. I will modify the max_unpooling layer based on graph pipeline now, the gradient check codes needs to be fixed as rebased now. Forward is fine for deconv.
@bhack The deconvolutional auto-encoder based on tiny_cnn could have an acceptable visualization, I use average unpooling rather than the max unpoooling using index produced by max pooling in deconvolution segmentation. The semantic segmentation have a dependency on the upcomming index-lead max unpoooling layer.

File: test/test_deconvolutional_layer.h [hunk 6]
```diff
@@ -90,5 +90,97 @@ TEST(deconvolutional, fprop) {
...
+TEST(deconvolutional, gradient_check) { // tanh - mse
+    network<sequential> nn;
+    nn << deconvolutional_layer<tan_h>(2, 2, 3, 1, 1);
+
+    vec_t a(4, 0.0);
+    label_t t = 3;
+
+    uniform_rand(a.begin(), a.end(), -1, 1);
+    nn.init_weight();
+    EXPECT_TRUE(nn.gradient_check<mse>(&a, &t, 1, 1e-4, GRAD_CHECK_ALL));
```
Comment (edgarriba) [inline test/test_deconvolutional_layer.h]: @Wangyida Try with float_t as float and double since there's a minor change of precision.

Comment (Wangyida) [inline test/test_deconvolutional_layer.h]: OK, I am rebasing now, these codes passed test as deconvolutional:[ PASSED ] in the older version.

**Comment by Wangyida** [comment]
@edgarriba @bhack Rebased on passed the tests.

**Comment by edgarriba** [comment]
it looks good to me

**Comment by Wangyida** [comment]
@edgarriba As the backend and some core functions changed since your PR 166 merged, I should modify the deconvolution and unpooling layer now. The visulization codes of 2 layers of coupled convolution-deconvolution anto encoder will produce an interesting reconstruction image in my example.

@mtamburrano@bhack The SqueezeNet based tiny model of deconvolution segmentation network is training now, it's not fine-tuned from VGG-16 model cause core micro structure of CNN is changed. 
I think there two main tasks now:

1. Modify the softmax loss to a pixel wise softmax loss used in the deconvolution semantic segmentation.
2. Turn the average unpooling to a coupled max pooling-unpooling pair for segmentation.

**Comment by edgarriba** [comment]
@Wangyida
try to adapt your deconvolutional models to new backend API.

Use this as an example: https://github.com/edgarriba/tiny-cnn/blob/feat/generic-computational-graph-device-abstraction/tiny_cnn/core/tiny_backend.h#L49

@bhack @nyanp 
Besides, I think that we could refactor a bit the tiny backend since the more operations we include the more number of lines will have this piece of code. Probably we can split tiny_backend in different kernels like `tiny_backend::conv_kernel()`  `tiny_backend::deconv_kernel()` and so on.

**Comment by nyanp** [comment]
@Wangyida 
Great! Would you like to add quick readme.md into your deconv examples? It's benefitical for everyone to understand what your examples do. 

@edgarriba 
I agree, as you are working in #192, we may want to split kernels per operations. Since all tests are passed, I think this PR is acceptable before using new backend API. This refactoring can be done by another PR :)

**Comment by Wangyida** [comment]
@nyanp Yes, I will add readme.md for core detail of forward and backward in my deconvolution codes, so other might have better solution on it because deconvolution has many SGD optimization method. And this .md file will also be used for a explanation on example.
@edgarriba I am implementing my original codes on to your backend style, I didn't pushed it as a new commit on this PR because some details are not proper and it might break the Travis online building. I will push another commit when they are neat.

**Comment by nyanp** [comment]
@Wangyida 
Ok, thank you very much! I'll merge this PR as soon as you'll push the doc.

**Comment by Wangyida** [comment]
@nyanp  Is this readme file OK? I see some more detailed description for the tiny_cnn related detail is already shown in mnist/readme.md:)

**Comment by nyanp** [comment]
@Wangyida
Ok, thank you very much :)
